ipam: Fix inconsistent update of CiliumNodes

pchaigno opened the pull request:
Currently, when the cilium-operator attaches new ENIs to a node, [we update the corresponding CiliumNode in two steps](https://github.com/cilium/cilium/blob/v1.12.0-rc2/pkg/ipam/node.go#L966-L1012): first the `.Status`, then the `.Spec`. That can result in an inconsistent state, where the CiliumNode `.Spec.IPAM.Pool` contains new IP addresses associated with the new ENI, while `.Status.ENI.ENIs` is still missing the ENI. This inconsistency manifests as a fatal:

    level=fatal msg="Error while creating daemon" error="Unable to allocate router IP for family ipv4: failed to associate IP 10.12.14.5 inside CiliumNode: unable to find ENI eni-9ab538c64feb9f59e" subsys=daemon

This inconsistency occurs because the following can happen:
1. cilium-operator attaches a new ENI to the CiliumNode.
2. Still at cilium-operator, `.Spec` is synced with kube-apiserver. The IP pool is updated with a new set of IP addresses and the new ENI.
3. The agent receives this half-updated CiliumNode. 
4. It allocates an IP address for the router from the pool of IPs attached to the new ENI, using `.Spec.IPAM.Pool`.
5. It fails because the new ENI is not listed in the `.Status.ENI.ENIs` of the CiliumNode object.
6. At cilium-operator, `.Status` is updated with the new ENI.

But wait, you said `.Status` is updated before `.Spec` in the function you linked? Yes, but we read the state to populate CiliumNode from two separate places (`n.ops.manager.instances` and `n.available`) in the `syncToAPIServer` function and we don't have anything to prevent having a half updated (one place only) state in the middle of the update function. We lock twice, once for each place, instead of once for the while CiliumNode update. So having a half updated state in the middle of the function would technically be the same as updating `.Spec` first and `.Status` second.

We can fix this by first creating a snapshot of the pool first, then write the .Status metadata (which may be more recent than the pool snapshot, which is safe, see comment in the source code of this patch), and then write the pool to .Spec. This ensures that the .Status is always updated before .Spec, but at the same time also ensures that .Status is still more recent than .Spec.

```release-note
Fix race condition leading to inconsistent CiliumNode that can cause the agent to fatal.
```

pchaigno committed:
ipam: Fix inconsistent update of CiliumNodes

Currently, when the cilium-operator attaches new ENIs to a node, we
update the corresponding CiliumNode in two steps: first the .Status,
then the .Spec [1]. That can result in an inconsistent state, where the
CiliumNode .Spec.IPAM.Pool contains new IP addresses associated with the
new ENI, while .Status.ENI.ENIs is still missing the ENI. This
inconsistency manisfests as a fatal:

    level=fatal msg="Error while creating daemon" error="Unable to allocate router IP for family ipv4: failed to associate IP 10.12.14.5 inside CiliumNode: unable to find ENI eni-9ab538c64feb9f59e" subsys=daemon

This inconsistency occurs because the following can happen:
1. cilium-operator attaches a new ENI to the CiliumNode.
2. Still at cilium-operator, .Spec is synced with kube-apiserver. The IP
   pool is updated with a new set of IP addresses and the new ENI.
3. The agent receives this half-updated CiliumNode.
4. It allocates an IP address for the router from the pool of IPs
   attached to the new ENI, using .Spec.IPAM.Pool.
5. It fails because the new ENI is not listed in the .Status.ENI.ENIs of
   the CiliumNode object.
6. At cilium-operator, .Status is updated with the new ENI.

But wait, you said .Status is updated before .Spec in the function you
linked? Yes, but we read the state to populate CiliumNode from two
separate places (n.ops.manager.instances and n.available) in the
syncToAPIServer function and we don't have anything to prevent having a
half updated (one place only) state in the middle of the update function.
We lock twice, once for each place, instead of once for the while
CiliumNode update. So having a half updated state in the middle of the
function would technically be the same as updating .Spec first and
.Status second.

We can fix this by first creating a snapshot of the pool first, then
write the .Status metadata (which may be more recent than the pool
snapshot, which is safe, see comment in the source code of this patch),
and then write the pool to .Spec. This ensures that the .Status is
always updated before .Spec, but at the same time also ensures that
.Status is still more recent than .Spec.

1 - https://github.com/cilium/cilium/blob/v1.12.0-rc2/pkg/ipam/node.go#L966-L1012

Co-authored-by: Sebastian Wicki <sebastian@isovalent.com>
Signed-off-by: Paul Chaignon <paul@cilium.io>
Signed-off-by: Sebastian Wicki <sebastian@isovalent.com>

christarazi commented on pkg/aws/eni/node.go:
```diff
@@ -596,9 +596,6 @@ var adviseOperatorFlagOnce sync.Once
 func (n *Node) GetMinimumAllocatableIPv4() int {
-	n.mutex.RLock()
-	defer n.mutex.RUnlock()
-
```
What happens to parallel access to these functions now? Do we assume that the lock is taken?

christarazi commented on pkg/ipam/node.go:
```diff
@@ -449,11 +449,9 @@ func (n *Node) releaseNeeded() (needed bool) {
 // Pool returns the IP allocation pool available to the node
 func (n *Node) Pool() (pool ipamTypes.AllocationMap) {
 	pool = ipamTypes.AllocationMap{}
-	n.mutex.RLock()
```
And this function as well

gandro committed:
ipam: Remove superfluous if statement

The `node.Spec.IPAM.Pool` value is always overwritten after the removed
`if` statement, so there is no need to initialize it.

Signed-off-by: Sebastian Wicki <sebastian@isovalent.com>

gandro commented on pkg/ipam/node.go:
```diff
@@ -963,6 +963,19 @@ func (n *Node) syncToAPIServer() (err error) {
+	// the pool after we updated the status, thereby creating a situation where
+	// the agent does not have the necessary IPAM information to use the newly
+	// added IP.
+	// When an IP is removed, this is also safe. IP release is done via
+	// handshake, where the agent will never use any IP where it has
+	// acknowledged the release handshake. Therefore, having an already
+	// released IP in the pool is fine, as the agent will ignore it.
```
@hemanthmalla Since you are more familiar with the release logic, could you confirm that this argument is sound?

hemanthmalla replied:
Yes, agent will ignore the IPs in the middle of a handshake for allocations. So it's safe. See https://github.com/DataDog/cilium/blob/b5fd71b9f873942764e90e41b39821ea48152832/pkg/ipam/crd.go#L564-L566

christarazi reviewed (approved):
Makes sense to me. Is this fixing https://github.com/cilium/cilium/issues/18366?

pchaigno reviewed (commented):
Looks good to me and cleaner than the first approach!
